pkg/elb/cleanup.go

```diff
@@ -0,0 +1,314 @@
+	}
+	if lb == nil {
+		// The load balancer wasn't found
+		return map[string]struct{}{}, nil
+	}
+
+	sgResponse, err := describeSecurityGroupsByID(ctx, ec2API, aws.StringValueSlice(lb.SecurityGroups))
+
+	if err != nil {
+		return nil, fmt.Errorf("error querying security groups for ELB %s: %s", loadBalancerName, err)
+	}
+
+	result := map[string]struct{}{}
+
+	for _, sg := range sgResponse.SecurityGroups {
+		sgID := aws.StringValue(sg.GroupId)
+
+		// FIXME(fons): AWS' CloudConfig accepts a global ELB security group, which shouldn't be deleted.
+		//              However, there doesn't seem to be a way to access the CloudConfiguration through the API Server.
+		//              Regardless, EKS doesn't expose the CloudConfiguration at the time of writing (which in
+		//              turn doesn't allow setting the ELB security group).
+		// if sgID == cfg.Global.ElbSecurityGroup {
+		//	//We don't want to delete a security group that was defined in the Cloud Configuration.
+		//	continue
+		// }
+
+		if sgID == "" {
+			continue
+		}
+
+		// Only delete security groups created by the Kubernetes Cloud provider
+		if !tagsIncludeClusterName(sg.Tags, clusterName) {
+			continue
+		}
+
+		result[sgID] = struct{}{}
+	}
+
+	return result, nil
+}
+
+func getLoadBalancerKind(service *corev1.Service) int {
+	// See https://github.com/kubernetes/kubernetes/blob/v1.12.6/pkg/cloudprovider/providers/aws/aws_loadbalancer.go#L51-L56
+	if service.Annotations[awsprovider.ServiceAnnotationLoadBalancerType] == "nlb" {
+		return loadBalancerKindNetwork
+	}
+	return loadBalancerKindClassic
+}
+
+func loadBalancerExists(ctx context.Context, ec2API ec2iface.EC2API, elbAPI elbiface.ELBAPI, elbv2API elbv2iface.ELBV2API, lb loadBalancer) (bool, error) {
+	exists, err := elbExists(ctx, elbAPI, elbv2API, lb.name, lb.kind)
+	if err != nil {
+		return false, err
+	}
+	if exists {
+		return true, nil
+	}
+
+	// Check whether all the security groups owned by the load balancer have also been deleted
+	// (they are deleted after the load balancer)
+	sgResponse, err := describeSecurityGroupsByID(ctx, ec2API, convertStringSetToSlice(lb.ownedSecurityGroupIDs))
+	if err != nil {
+		return false, err
+	}
+	return len(sgResponse.SecurityGroups) != 0, nil
+}
+
+func elbExists(ctx context.Context, elbAPI elbiface.ELBAPI, elbv2API elbv2iface.ELBV2API,
+	name string, kind int) (bool, error) {
+	if kind == loadBalancerKindNetwork {
+		return elbV2Exists(ctx, elbv2API, name)
+	}
+	desc, err := describeClassicLoadBalancer(ctx, elbAPI, name)
+	return desc != nil, err
+}
+
+func describeClassicLoadBalancer(ctx context.Context, elbAPI elbiface.ELBAPI,
+	name string) (*elb.LoadBalancerDescription, error) {
+
+	request := &elb.DescribeLoadBalancersInput{
+		LoadBalancerNames: []*string{&name},
+	}
+
+	response, err := elbAPI.DescribeLoadBalancersWithContext(ctx, request)
+	if err != nil {
+		if awsError, ok := err.(awserr.Error); ok {
+			if awsError.Code() == "LoadBalancerNotFound" {
```

Comment:
They have `elbAPI.ErrCodeAccessPointNotFoundException`, but that reads rather weird, so indeed let's use a string.

Reply:
I got this from https://github.com/kubernetes/legacy-cloud-providers/blob/release-1.15/aws/aws.go#L2720-L2723

`ErrCodeAccessPointNotFoundException` doesn't exist in the SDK https://github.com/aws/aws-sdk-go/search?q=ErrCodeAccessPointNotFoundException&unscoped_q=ErrCodeAccessPointNotFoundException

Reply:
Actually, it does exist, I will use that